src/paxos.rs

```diff
@@ -189,6 +191,72 @@ where
         )
     }
 
+    /// Initiates the garbage collection process using the provided index.
+    /// # Arguments
+    /// * `index` - Deletes all entries up to [`index`]
+    pub fn garbage_collect_by_index(&mut self, index: u64) {
+        self.gc_propose(Some(index));
+    }
+
+    /// Initiates the garbage collection process using the minimum last accepted sequence as the [`index`].
+    pub fn garbage_collect(&mut self) {
+        self.gc_propose(None);
+    }
+
+    /// Return garbage collection index from storage.
+    pub fn get_gc_from_storage(&self) -> u64 {
+        self.storage.get_gc_idx()
+    }
+
+    fn gc_propose(&mut self, index: Option<u64>) {
+        match self.state {
+            (Role::Leader, _) => {}
+            _ => self.forward_gc_proposal(index),
+        }
+    }
+
+    fn gc_leader(&mut self, index: Option<u64>) {
+        let mut idx;
+        if index.is_some() {
+            idx = self.las.iter().min().cloned();
+            if (idx.is_none()) | (idx.unwrap() < index.unwrap()) {
+                return;
+            }
+            idx = index;
+        } else {
+            idx = self.las.iter().min().cloned();
+            if idx.is_none() {
+                return;
+            }
+        }
+
+        for pid in &self.peers {
+            self.outgoing
+                .push(Message::with(self.pid, *pid, PaxosMsg::GarbageCollect(idx)));
+        }
+
+        self.gc(idx);
+    }
+
+    fn gc(&mut self, index: Option<u64>) {
+        if index.is_none() | (self.storage.get_decided_len() < index.unwrap()) {
+            crit!(
+                self.logger,
+                "Received invalid garbage collection index! index: {:?}, prev_ld {}",
+                index,
+                self.prev_ld
+            );
+            return;
+        }
+
+        trace!(self.logger, "Garbage Collection index: {:?}", index);
+
+        self.cached_gc_index = index.unwrap();
+
+        self.storage.set_gc_idx(index.unwrap());
+        self.storage.garbage_collect(index.unwrap());
```

Comment:
we **must** call `self.storage.garbage_collect()` first. We don't want to first set the index and then crash and when we recover we will have the updated gc_index but the entries are still in the log.